pinakes/common/auth/keycloak_django/authentication.py

```diff
@@ -0,0 +1,114 @@
+from typing import Any, Optional, Tuple
+
+import requests
+from django.contrib.auth.models import AbstractUser
+from django.core.exceptions import ObjectDoesNotExist
+from django.utils.translation import gettext_lazy as _
+from rest_framework import status
+from rest_framework.authentication import (
+    BaseAuthentication,
+    SessionAuthentication,
+)
+from rest_framework.exceptions import AuthenticationFailed
+from rest_framework.request import Request
+from social_django.models import AbstractUserSocialAuth
+from social_django.utils import load_strategy
+
+from pinakes.common.auth.keycloak_oidc import KeycloakOpenIdConnect
+
+
+UserTokenPair = Tuple[AbstractUser, Any]
+
+
+class KeycloakSessionAuthentication(SessionAuthentication):
+    social_auth_provider = "keycloak-oidc"
+
+    def authenticate(self, request: Request) -> Optional[UserTokenPair]:
+        auth = super().authenticate(request)
+        if auth is None:
+            return None
+
+        user = auth[0]
+
+        keycloak_user = self.get_social_user(user)
+        strategy = load_strategy(request)
+        try:
+            token = keycloak_user.get_access_token(strategy)
+        except requests.HTTPError as exc:
+            if exc.response.status_code == status.HTTP_400_BAD_REQUEST:
+                raise AuthenticationFailed(_("Access token expired"))
+            raise
+
+        if token is None:
+            raise AuthenticationFailed(_("Access token is missing"))
+
+        return user, token
+
+    def get_social_user(
+        self, user: AbstractUser
+    ) -> Optional[AbstractUserSocialAuth]:
+        social_auth = getattr(user, "social_auth", None)
+        if social_auth is None:
+            return None
+
+        try:
+            return social_auth.get(provider=self.social_auth_provider)
+        except ObjectDoesNotExist:
+            return None
+
+
+class BaseKeycloakBearerAuthentication(BaseAuthentication):
+    auth_scheme = "Bearer"
+    auth_backend = KeycloakOpenIdConnect
+
```

Comment:
@cutwater do we need to define the authenticate method here. https://github.com/encode/django-rest-framework/blob/7e4e6d207070d50736827a281b5cb70eb161b782/rest_framework/authentication.py#L38 since it is inheriting from BaseAuthentication.